File: Javascript/Arrays/containsDuplicate.js
var containsDuplicates = nums => {

   for(let i = 0; i < nums.length - 1; i++){
    for(let j = i + 1; j < nums.length; j++){
        if(nums[i] === nums[j]){
            return true;
        }
    }

   } return false;
   
   /* Time complexity : O(n^2)
    This is the brute force solution which is far from optimal.
    Explanation :
        i = current element , j = adjacent element 
        we compare the i to its adjancent to know if there are duplicates
        */
}

// Optimal - Solution
const containsDuplicates2 = function(nums) {
    const set = new Set(nums);

    return set.size !== nums.length 

    // Time complexity : O(n) - because of the set method which iterates over the array
}



// TO run the function:
// containsDuplicates([1,2,4,3,5,6,7]);

File: Python/Arrays/containsDuplicate.py
def containsDuplicates(nums):
    # Making sure that the values are different or the same
    noDuplicates = set(nums)
    
    # If the length of them are different it means that we have found duplicates
    if(len(noDuplicates) != len(nums)):
        return True 
    # if that's not the case we know that there's no duplicates
    else:
        return false



""" Time complexity : O(n) , where n is the length of the input list nums.

This is because the set operation used to remove duplciates has a time complexity of O(n)
as it requires iterating over the entire list to check for duplicates, and the subsquent 
comparison of the length of the original list and the set also has a time compleixty of O(n)

Space Complexity :  O(n) , as the set operation creates a new set with potentially all n elements
from the input list. However, the space complexity could be reduced to O(K), where k is the number of unique elements
in the input list, if we only added elements to the set until we found a duplicate

"""

# Improved version 

def containsDuplicates2(nums):

    seen = set()
    for num in nums:
        if num in seen:
            return True
        seen.add(num)
    return False 

"""This solution is better because k represents the number of unique elements in the input list. 
This is because the algorithm uses a set to keep track of only the unique elements seen so far, 
and the size of this set will be at most equal to the number of unique elements in the input list """
